eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Utxo.hs

```diff
@@ -611,29 +610,36 @@ validateMaxTxSizeUTxO pp tx =
     txSize = tx ^. sizeTxF
 
 updateUTxOState ::
-  EraTxBody era =>
+  (EraTxBody era, Monad m) =>
   PParams era ->
   UTxOState era ->
   TxBody era ->
-  Coin ->
+  CertState era ->
   GovState era ->
-  UTxOState era
-updateUTxOState pp UTxOState {utxosUtxo, utxosDeposited, utxosFees, utxosStakeDistr, utxosDonation} txb depositChange govState =
-  let UTxO utxo = utxosUtxo
-      !utxoAdd = txouts txb -- These will be inserted into the UTxO
+  (Coin -> m ()) ->
+  m (UTxOState era)
+updateUTxOState pp utxos txBody certState govState depositChangeEvent = do
+  let UTxOState {utxosUtxo, utxosDeposited, utxosFees, utxosStakeDistr, utxosDonation} = utxos
+      UTxO utxo = utxosUtxo
+      !utxoAdd = txouts txBody -- These will be inserted into the UTxO
       {- utxoDel  = txins txb ◁ utxo -}
-      !(utxoWithout, utxoDel) = extractKeys utxo (txb ^. inputsTxBodyL)
+      !(utxoWithout, utxoDel) = extractKeys utxo (txBody ^. inputsTxBodyL)
       {- newUTxO = (txins txb ⋪ utxo) ∪ outs txb -}
       newUTxO = utxoWithout `Map.union` unUTxO utxoAdd
       newIncStakeDistro = updateStakeDistribution pp utxosStakeDistr (UTxO utxoDel) utxoAdd
-   in UTxOState
-        { utxosUtxo = UTxO newUTxO
-        , utxosDeposited = utxosDeposited <> depositChange
-        , utxosFees = utxosFees <> txb ^. feeTxBodyL
-        , utxosGovState = govState
-        , utxosStakeDistr = newIncStakeDistro
-        , utxosDonation = utxosDonation
-        }
+      totalRefunds = certsTotalRefundsTxBody pp certState txBody
+      totalDeposits = certsTotalDepositsTxBody pp certState txBody
+      depositChange = totalDeposits <-> totalRefunds
+  depositChangeEvent depositChange
+  pure $!
+    UTxOState
+      { utxosUtxo = UTxO newUTxO
+      , utxosDeposited = utxosDeposited <> depositChange
```

Comment:
Moving the computation of deposits and refunds to the function that makes the changes to the UTxOState make sense, it is a natural extension to 'smartUTxOState, in that it automatically makes the invariants true. Unlike smartUTxOState it is NOT really expensive.  But to someone who is just reading this PR, or someone who is just using unpdateUTxOState a lot of implicit changes happen that user is not aware of. We should add some comments to the defintion of updateUTxOState to make this clear.

Reply:
I am certainly on board with adding some haddock explaining this function.

Reply:
Done